fiat/coindesk_api.go

```diff
@@ -0,0 +1,117 @@
+package fiat
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"time"
+
+	"github.com/shopspring/decimal"
+
+	"github.com/lightninglabs/faraday/utils"
+)
+
+const (
+	// coinDeskHistoryAPI is the endpoint we hit for historical price data.
+	coinDeskHistoryAPI = "https://api.coindesk.com/v1/bpi/historical/close.json"
+
+	// coinDeskTimeFormat is the date format used by coindesk.
+	coinDeskTimeFormat = "2006-01-02"
+)
+
+// coinDeskAPI implements the PriceAPIBackend interface.
+type coinDeskAPI struct{}
+
+type coinDeskResponse struct {
+	Data map[string]float64 `json:"bpi"`
+}
+
+// queryCoinDesk constructs and sends a request to coindesk to query historical
+// price information.
+func queryCoinDesk(start, end time.Time) ([]byte, error) {
+	queryURL := fmt.Sprintf("%v?start=%v&end=%v",
+		coinDeskHistoryAPI, start.Format(coinDeskTimeFormat),
+		end.Format(coinDeskTimeFormat))
+
+	log.Debugf("coindesk url: %v", queryURL)
+
+	// Query the http endpoint with the url provided
+	// #nosec G107
+	response, err := http.Get(queryURL)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	return ioutil.ReadAll(response.Body)
+}
+
+// parseCoinDeskData parses http response data from coindesk into USDPrice
+// structs.
+func parseCoinDeskData(data []byte) ([]*USDPrice, error) {
+	var priceEntries coinDeskResponse
+	if err := json.Unmarshal(data, &priceEntries); err != nil {
+		return nil, err
+	}
+
+	// nolint: prealloc
+	var usdRecords []*USDPrice
+
+	for date, price := range priceEntries.Data {
+		timestamp, err := time.Parse(coinDeskTimeFormat, date)
+		if err != nil {
+			return nil, err
+		}
+
+		usdRecords = append(usdRecords, &USDPrice{
+			Timestamp: timestamp,
+			Price:     decimal.NewFromFloat(price),
+		})
+	}
+
+	return usdRecords, nil
+}
+
+// GetPrices retrieves price information from coindesks's api for the given
+// time range.
+func (c *coinDeskAPI) GetPrices(ctx context.Context, start,
```

Comment:
Time range validation and sort by timestamp is going to be common to all fiat backends, so could we cut the duplication out?

Thinking something like this:
```
type FiatSource struct{
    impl fiatBackend
}

func (f *FiatSource) GetPrices(){
   utils.ValidateTimeRange
   
   records, err := f.impl.rawPriceData()
    
   sort(records)
}
```

Where `fiatBackend` would be coincap/coindesk implementation that just fetches data. You'd need to rename the coincap `GetPrices` and move validation out, but then we can just return a `FiatSource` and we don't even need to export the interface.